### PR TITLE
Output Closure annotations when --closure

### PR DESCRIPTION
**Pull request by ochafik**

### CODE REVIEW

**Comment by jmesserly** [comment]
Really nice work. CC @vsmenon and @leafpetersen for additional thoughts.

Main thing I'm wondering is what to do about precompiled SDK JS files. I don't think we want those to use `--closure` but not sure what alternative to suggest. Related to https://github.com/dart-lang/dev_compiler/issues/257 (compiling against a different SDK than we run with) and https://github.com/dart-lang/dev_compiler/issues/223 (incremental compile). The precompiled SDK is kind of a hack to avoid SDK build time, but it causes problems like this one.

I guess we have a related problem with hand coded libraries. They turned out to be a lot bigger than I was expecting. Since we haven't used the `JS` builtin style of dart2js, we don't have ability to apply compiler transforms to them (like 6->5 lowering, closure annotations, minify, etc.)

Anyway, maybe we can discuss this further here or have a design meeting offline. Cheers, - John

**Comment by ochafik** [comment]
Hey John, I've added a bit of tests + dropped the `--closure` flag from build_sdk.sh. Instead, for now, people who want to play with the closure backend will have to rebuild the sdk with `./tool/build_sdk.sh --closure` (when things are production-strength, we can talk again about that smart 1-time-cost + caching).

I don't have an answer for hard coded libraries, except that... what prevents us from writing them in Dart? (maybe some annotation could trigger a dart:js pass-through mode, which would allow to express near-complete JS semantics?).

**Comment by jmesserly** [comment]
> I don't have an answer for hard coded libraries, except that... what prevents us from writing them in Dart? (maybe some annotation could trigger a dart:js pass-through mode, which would allow to express near-complete JS semantics?).

CC @vsmenon @leafpetersen for thoughts?

The dart2js style is to have them in a private `dart:` library and with "JS" snippets inside method/function bodies.

The context here: if we have multiple output modes (ES6, Closure, TypeScript, maybe something else in the future? also different JS module systems) it might be easier to maintain our runtime helpers if they're in a Dart library vs a .js file.

File: lib/src/closure/closure_annotation.dart [hunk 1]
```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2012, the Dart project authors.  Please see the AUTHORS file
```
Comment (jmesserly) [inline lib/src/closure/closure_annotation.dart]: 2015 in the various new files?

**Comment by jmesserly** [comment]
this LGTM ... @leafpetersen would you be interested in reviewing the type annotation code?

**Comment by jmesserly** [comment]
aside: this is really great looking code! I really like how it reads and is factored

File: lib/src/closure/closure_annotation.dart (hunk 1 above)
Comment (ochafik) [inline lib/src/closure/closure_annotation.dart]: Oops, done.

**Comment by ochafik** [comment]
Thanks a lot John!

File: lib/src/closure/closure_codegen.dart [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+import 'package:analyzer/analyzer.dart' show ParameterKind;
+import 'package:analyzer/src/generated/element.dart';
+
+import 'closure_annotation.dart';
+import 'closure_type.dart';
+import 'package:analyzer/src/generated/resolver.dart' show TypeProvider;
+
+/// Mixin that can generate [ClosureAnnotation]s for Dart elements and types.
+abstract class ClosureCodegen {
```
Comment (vsmenon) [inline lib/src/closure/closure_codegen.dart]: Minor nit: ClosureCodegen -> ClosureAnnotator ?

File: lib/src/closure/closure_codegen.dart [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        return new ClosureType.function();
+      default:
+        return new ClosureType.type(getQualifiedName(classElement));
+    }
+  }
+
+  ClosureType _closureTypeForDartType(DartType type) {
+    // TODO(ochafik): Work out the practical difference between all (*) and unknown (?).
+    if (type == null) return new ClosureType.unknown();
+    if (type.isDynamic) return new ClosureType.all();
```
Comment (vsmenon) [inline lib/src/closure/closure_codegen.dart]: Maybe a TODO here to work out if dynamic needs to ever be mapped to bottom?  Is there such a type in Closure?

File: lib/src/closure/closure_codegen.dart (hunk 2 above)
Comment (ochafik) [inline lib/src/closure/closure_codegen.dart]: Done.

File: lib/src/closure/closure_codegen.dart (hunk 3 above)
Comment (ochafik) [inline lib/src/closure/closure_codegen.dart]: Hehe, so I've [played a bit with function params](https://gist.github.com/ochafik/d8f701557f5bcab68940) and downcasts in Closure, following @leafpetersen's comments, and looks like we probably want to map pretty much everything weirdly-typed (dynamic, bottom) to unknown `?` (see [doc](https://developers.google.com/closure/compiler/docs/js-for-compiler#types)).

**Comment by leafpetersen** [comment]
LGTM.

**Pull request merged**